File: java/Binary_Search/Day8/Nth_Root.java
package Day8;
import java.util.*;

public class Nth_Root {
    // Function to calculate x^n (power)...
    // public static long power(long x, long n) {
    //     long ans = 1;
    //     for (int i = 1; i <= n; i++) {
    //         ans *= x;
    //     }
    //     return ans;
    // }



    // Function to calculate mid^n safely (to avoid overflow)...
    public static long powerExpo(long mid, long n, long m) {
        // return 1 if == m...
        // return 0 if < m...
        // return 2 if > m...
        long ans = 1;
        for (int i = 1; i <= n; i++) {
            ans = ans * mid;
            if (ans > m) return 2;
        }
        if (ans == m) return 1;

        return 0;
    }


    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("No of Nth Root:- ");
        long n = sc.nextLong();

        System.out.print("Number:- ");
        long m = sc.nextLong();

        long ans = -1;


        // ---------- Linear Search ----------
        // for (long i = 1; i <= m; i++) {
        //     long val = power(i, n);

        //     if (val == m) {
        //         ans = i;
        //         break;
        //     } else if (val > m) {
        //         break;
        //     }
        // }

        // System.out.println("Nth Root = " + ans);





        // ---------- Binary Search ----------
        long low = 1, high = m;

        while (low <= high) {
            long mid = (low + high) / 2;
            long midN = powerExpo(mid, n, m);

            if (midN == 1) {
                ans = mid;
                break;
            }
            else if (midN == 0) low = mid + 1;
            else high = mid - 1;
        }
        
        System.out.println("Nth Root = " + ans);

        sc.close();
    }
}

File: python/Binary_Search/day8/nth_root.py
# ---------- Linear Search ----------
# # Function to calculate x^n (power)...
# def power(x, n):
#     ans = 1
#     for i in range(n):
#         ans *= x
    
#     return ans


# # main part...
# n = int(input("No of Nth Root:- "))
# m = int(input("Number:- "))

# ans = -1   # default (not found)...

# for i in range(1, m + 1):
#     val = power(i, n)
#     if val == m:
#         ans = i
#         break
#     elif val > m:
#         break

# print("Nth Root =", ans)








# ---------- Binary Search ----------
# return 1 if == m...
# return 0 if < m...
# return 2 if > m...
def powerExpo(mid, n, m):
    ans = 1
    for i in range(n):
        ans = ans * mid
        if ans > m:
            return 2
    
    if ans == m:
        return 1
    
    return 0





# main part...
n = int(input("No of Nth Root:- "))
m = int(input("Number:- "))

ans = -1   # default (not found)...

low = 1
high = m

while low <= high:
    mid = (low + high) // 2
    midN = powerExpo(mid, n, m)

    if midN == 1:
        ans = mid
        break

    elif midN == 0:
        low = mid + 1
    
    else:
        high = mid - 1

print("Nth Root =", ans)